Project 9: Implemented API

yuranest opened the pull request:
- Fixed API endpoints for likes/unlikes.
- Ensured `_id` exists before creating a card.
- Fixed errors with deleting/liking cards.
- Improved form validation and UX.

yuranest committed:
avatar button add

yuranest committed:
Save... set

yuranest committed:
Avatar change

yuranest committed:
Like fix

yuranest committed:
Like fix

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Yuriy, 

unfortunately, I have to reject the project without review. If you need any assistance, our tutors are here to help:


* Likes don't work in the initial cards https://snipboard.io/cYo0Jb.jpg 
* I can't delete initial cards https://snipboard.io/iagjbO.jpg
* When I edit the profile data, the `job` doesn't change 
* When I add new cards, and reload the page, all my new cards disappear

yuranest committed:
fix / Likes / cards / deled / profil

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Yuriy, 

unfortunately, I have to reject the project without review. If you need any assistance, our tutors are here to help:


* The project starts without the content https://snipboard.io/Pdc16s.jpg  . Please, check your webpack settings

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Yuriy, 

unfortunately, I have to reject the project without review. If you need any assistance, our tutors are here to help:


* The project starts without the content https://snipboard.io/Pdc16s.jpg  . Please, check your webpack settings
* You need to send a link for your **opened** pull request with the code of this stage rather than a link for your **branch**

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Yuriy, 

unfortunately, I have to reject the project without review. If you need any assistance, our tutors are here to help:


* The `pull request` doesn't contain any changes after the previous rejection https://snipboard.io/ULlRkY.jpg . The issues are still the same

yuranest commented:
Pull ?

yuranest committed:
fix / Likes / cards / deled / profil

gennady-bars commented on webpack.config.js:
```diff
@@ -12,7 +12,7 @@ module.exports = {
     path: path.resolve(__dirname, "dist"), // Output directory for bundled files
     filename: "scripts/[name].[contenthash].js", // JS file naming pattern with content hash
-    publicPath: "/se_project_aroundtheus/", // ✅ Corrected for GitHub Pages "/se_project_aroundtheus/"
+    publicPath: "/se_project_aroundtheus/", // Corrected for GitHub Pages "/se_project_aroundtheus/"
```
(**Please, expand the general comment ↓**)

Dear Yuriy, 

unfortunately, I have to reject the project without review. If you need any assistance, our tutors are here to help:


* The project starts without the content https://snipboard.io/Pdc16s.jpg  . Please, check your webpack settings

yuranest replied:
https://github.com/yuranest/se_project_aroundtheus/pull/2

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Yuriy, 

unfortunately, I have to reject the project without review. If you need any assistance, our tutors are here to help:


* The project starts without the content https://snipboard.io/Pdc16s.jpg  . Please, check your webpack settings

yuranest committed:
webpack fix /

gennady-bars commented on src/components/Api.js:
```diff
@@ -0,0 +1,62 @@
+  }
+
+  // Generic method to handle API requests
+  async _request(url, method, body = null) {
```
Nice `_request` method

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -2,37 +2,61 @@ import Popup from "./Popup.js";
+    this._formValidator = formValidator;
+  }
+
+  _showLoading(isLoading) {
+    this._submitButton.textContent = isLoading
+      ? "Saving..."
+      : this._defaultButtonText;
+    this._submitButton.disabled = isLoading;
```
Nice `_showLoading` method

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -2,37 +2,61 @@ import Popup from "./Popup.js";
     });
     return inputValues;
   }
 
```
**Could be improved**

You can make ` setInputValues` in class `PopupWithForm`, which can insert data into inputs:

```js

  setInputValues(data) {
    this._inputList.forEach((input) => {
      // Here you insert the `value` by the `name` of the input
      input.value = data[input.name];
    });
  }



```

So, this way you’ll not have to search the inputs of the **profile** in `index.js`

gennady-bars commented on src/components/PopupWithForm.js:
```diff
@@ -2,37 +2,61 @@ import Popup from "./Popup.js";
+      }
     });
   }
 
```
**COULD BE IMPROVED**

You can make a special method for getting a form element from `PopupWithForm` to avoid wasting resources on searching it in the DOM in `index.js` , because you’ve already found it in the constructor:



```js

  getForm() {
    return this._form
  }


```

gennady-bars commented on src/pages/index.js:
```diff
@@ -4,81 +4,191 @@ import PopupWithForm from "../components/PopupWithForm.js";
+        });
+        document.forms["profile-form"].reset();
+      })
+      .catch((err) => console.error("Error updating user info:", err));
```
You don't need `catch` in `index.js` if you have it inside `PopupWithForm` at the very end of the promise chain

**Please, correct it everywhere in the code**

gennady-bars commented on src/pages/index.js:
```diff
@@ -4,81 +4,191 @@ import PopupWithForm from "../components/PopupWithForm.js";
+    return api
+      .updateUserAvatar({ avatar: data.avatar })
+      .then((res) => {
+        document.querySelector(".profile__avatar").src = res.avatar;
```
Only methods of `UserInfo` should set/get the profile data (including the **avatar**) 


You should use `userInfo.setUserInfo` here for setting the avatar

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Yuriy!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

**Parts I really liked in your project:**

* Good job that you **close** modals only in blocks `then`
* Good job that you `catch` possible errors at the end of server requests.
* Good job that you return the button text in `finally`
* Nice `_request` method

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code and the comments below:
* You don't need `catch` in `index.js` if you have it inside `PopupWithForm` at the very end of the promise chain
* Only methods of `UserInfo` should set/get the profile data (including the **avatar**) 



There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

yuranest committed:
Refactored UserInfo usage and removed redundant catch statements

yuranest commented:
- Refactored UserInfo to manage avatar updates correctly.
- Removed unnecessary `.catch()` in `index.js` (already handled in PopupWithForm).
- Ensured `setInputValues()` is used in PopupWithForm.
- Improved API request handling and validation.

gennady-bars commented on src/pages/index.js:
```diff
@@ -4,90 +4,185 @@ import PopupWithForm from "../components/PopupWithForm.js";
+document.querySelector(".avatar__edit-button").addEventListener("click", () => {
+  avatarPopup.open();
+});
+
+// ✅ Delete confirmation popup
+let cardToDelete = null;
+let cardIdToDelete = null;
+
+const deletePopup = new PopupWithForm(".modal_type_delete", () => {
+  if (cardToDelete && cardIdToDelete) {
+    return api.deleteCard(cardIdToDelete).then(() => {
+      cardToDelete.deleteCard();
+      cardToDelete = null;
+      cardIdToDelete = null;
+      deletePopup.close();
+    });
+  }
+  return Promise.reject("No card selected for deletion");
+});
+deletePopup.setEventListeners();
+
+// ✅ Function to open delete popup
+function openDeletePopup(cardId, cardInstance) {
+  if (!cardId) {
+    console.error("Error: Trying to delete a card without an ID");
+    return;
+  }
+  cardIdToDelete = cardId;
+  cardToDelete = cardInstance;
+  deletePopup.open();
+}
+
+// ✅ Event listeners for profile popups
 document
   .querySelector(".profile__edit-button")
   .addEventListener("click", () => {
     const userData = userInfo.getUserInfo();
-    document.forms["profile-form"].name.value = userData.name;
-    document.forms["profile-form"].description.value = userData.job;
+    profilePopup.setInputValues(userData);
```
* There is no `job` text in the `job` input when I open the `profile` modal https://snipboard.io/Na1qPt.jpg . Please, fill it correctly

You should update the object with `description` because this value is in the `name` attribute of the `job` input

```js

    const userData = userInfo.getUserInfo();
    userData.description = userData.job;  // update the object
    profilePopup.setInputValues(userData);
```

gennady-bars reviewed (commented):
Everything is almost fine, but:


* There is no `job` text in the `job` input when I open the `profile` modal https://snipboard.io/Na1qPt.jpg . Please, fill it correctly

yuranest committed:
input placeholder fix

yuranest committed:
input placeholder fix

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!
